### PR TITLE
[Extensions] Moving ExtensionsRequest to use Protobuf

### PR DESCRIPTION
**Pull request by saratvemulapalli**
### Description
As part of https://github.com/opensearch-project/opensearch-sdk-java/issues/337. Moving the first API to protobuf to have seamless version support. 

### Issues Resolved
Part of #https://github.com/opensearch-project/opensearch-sdk-java/issues/642
Also first introduction of protobuf #6844 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by saratvemulapalli** [comment]
```
FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':distribution:docker:buildDockerImage'.
> A failure occurred while executing org.opensearch.gradle.docker.DockerBuildTask$DockerBuildAction
   > Failed to pull Docker **** image [centos:8], all attempts failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
org.gradle.api.GradleException: Reaper process failed. Check log at /var/jenkins/workspace/gradle-check/search/.gradle/reaper/build-14854/error.log for details
> Reaper process failed. Check log at /var/jenkins/workspace/gradle-check/search/.gradle/reaper/build-14854/error.log for details

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
==============================================================================

```

File: server/src/main/java/org/opensearch/extensions/proto/ExtensionRequestOuterClass.java [hunk 1]
```diff
@@ -0,0 +1,956 @@
...
+
+// Generated by the protocol buffer compiler.  DO NOT EDIT!
+// source: server/src/main/java/org/opensearch/extensions/proto/ExtensionRequest.proto
+
+package org.opensearch.extensions.proto;
+
+/**
+ - * ExtensionRequest class autogenerated from ExtensionRequest.proto
+ - */
+public final class ExtensionRequestOuterClass {
```
Comment (reta) [inline server/src/main/java/org/opensearch/extensions/proto/ExtensionRequestOuterClass.java:20]: Hm ... this class is auto generated, shouldn't we generate it during the build (like the typical code generation flows)? (https://github.com/google/protobuf-gradle-plugin)

Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/proto/ExtensionRequestOuterClass.java:20]: This is nice, thanks @reta let me try getting this integrated.

File: server/src/main/java/org/opensearch/extensions/ExtensionsManager.java [hunk 2]
```diff
@@ -89,32 +89,13 @@ public class ExtensionsManager {
     public static final String REQUEST_EXTENSION_REGISTER_CUSTOM_SETTINGS = "internal:discovery/registercustomsettings";
     public static final String REQUEST_EXTENSION_REGISTER_REST_ACTIONS = "internal:discovery/registerrestactions";
     public static final String REQUEST_EXTENSION_REGISTER_TRANSPORT_ACTIONS = "internal:discovery/registertransportactions";
-    public static final String REQUEST_OPENSEARCH_PARSE_NAMED_WRITEABLE = "internal:discovery/parsenamedwriteable";
```
[the remark this answers is not in the document]
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:92]: Yup we dont use this anymore. 
cc: @ryanbogan

**Comment by saratvemulapalli** [comment]
@reta I was able to get plugin integrated and it works. 
Looks like publishing tasks are failing because of javaDoc errors. I'll head to that next.

```
extension-api-opensearch % ./gradlew publishToMavenLocal
Maven publication 'pluginMaven' pom metadata warnings (silence with 'suppressPomMetadataWarningsFor(variant)'):
  - Variant testFixturesApiElements:
      - Declares capability org.opensearch.gradle:build-tools-test-fixtures:3.0.0-SNAPSHOT which cannot be mapped to Maven
  - Variant testFixturesRuntimeElements:
      - Declares capability org.opensearch.gradle:build-tools-test-fixtures:3.0.0-SNAPSHOT which cannot be mapped to Maven
These issues indicate information that is lost in the published 'pom' metadata file, which may be an issue if the published library is consumed by an old Gradle version or Apache Maven.
The 'module' metadata file, which is used by Gradle 6+ is not affected.
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 8.0.2
  OS Info               : Mac OS X 10.16 (x86_64)
  JDK Version           : 14 (Oracle JDK)
  JAVA_HOME             : /Library/Java/JavaVirtualMachines/jdk-14.0.2.jdk/Contents/Home
  Random Testing Seed   : C28A2640421F3020
  In FIPS 140 mode      : false
=======================================

> Task :build-tools:publishPluginMavenPublicationToMavenLocal
Multiple publications with coordinates 'org.opensearch.gradle:build-tools:3.0.0-SNAPSHOT' are published to repository 'mavenLocal'. The publications 'nebula' in project ':build-tools' and 'pluginMaven' in project ':build-tools' will overwrite each other!

> Task :server:missingJavadoc FAILED
/Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:16: error: package org.opensearch.extensions.proto does not exist
import org.opensearch.extensions.proto.ExtensionRequestProto;
                                      ^
/Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:29: error: package ExtensionRequestProto does not exist
    private final ExtensionRequestProto.ExtensionRequest request;
                                       ^
/Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:31: error: package ExtensionRequestProto does not exist
    public ExtensionRequest(ExtensionRequestProto.RequestType requestType) {
                                                 ^
/Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:35: error: package ExtensionRequestProto does not exist
    public ExtensionRequest(ExtensionRequestProto.RequestType requestType, @Nullable String uniqueId) {
                                                 ^
/Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:54: error: package ExtensionRequestProto does not exist
    public ExtensionRequestProto.RequestType getRequestType() {
                                ^
5 errors

FAILURE: Build failed with an exception.

* Where:
Script '/Users/sarat/Desktop/workspace/Code/extension-api-opensearch/gradle/missing-javadoc.gradle' line: 356

* What went wrong:
Execution failed for task ':server:missingJavadoc'.
> Javadoc generation failed for :server,
    Options file at: /Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/build/tmp/missingJavadoc/javadoc-options.txt
    Command output at: /Users/sarat/Desktop/workspace/Code/extension-api-opensearch/server/build/tmp/missingJavadoc/javadoc-output.txt

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

BUILD FAILED in 26s
644 actionable tasks: 193 executed, 451 up-to-date
```

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 3]
```diff
@@ -26,54 +26,54 @@
  */
 public class ExtensionRequest extends TransportRequest {
```
Comment (peternied) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:27]: Can we have the generated protobuf class generate the transport request required methods while its at it?  I'd love to get rid of the error prone equals/hash/tostring also

**Comment by saratvemulapalli** [comment]
Still failing with Docker problems. 
@peterzhuamazon 

> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/13774/
> * **CommitID:** [e926b8d](https://github.com/opensearch-project/OpenSearch/commit/e926b8d3bdf778337e5ea4f2687a8c610d498b3c)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

File: plugins/repository-hdfs/build.gradle [hunk 4]
```diff
@@ -116,14 +115,7 @@ tasks.named("dependencyLicenses").configure {
...
-    'com.google.protobuf.MessageSchema',
-    'com.google.protobuf.UnsafeUtil',
-    'com.google.protobuf.UnsafeUtil$1',
-    'com.google.protobuf.UnsafeUtil$Android32MemoryAccessor',
-    'com.google.protobuf.UnsafeUtil$Android64MemoryAccessor',
-    'com.google.protobuf.UnsafeUtil$JvmMemoryAccessor',
-    'com.google.protobuf.UnsafeUtil$MemoryAccessor'
+
   )
 }
```
Comment (owaiskazi19) [inline plugins/repository-hdfs/build.gradle:120]: Do we need this anymore?

File: server/build.gradle [hunk 5]
```diff
@@ -30,6 +30,10 @@
 
 import org.opensearch.gradle.info.BuildParams
 
+plugins {
+  id('com.google.protobuf') version '0.8.19'
```
Comment (owaiskazi19) [inline server/build.gradle:34]: `com.google.protobuf` has `0.9.2` latest version. Also, can we add the version in version.properties and fetch it from there?

File: server/build.gradle [hunk 6]
```diff
@@ -136,6 +147,9 @@ dependencies {
   // jna
   api "net.java.dev.jna:jna:${versions.jna}"
 
+  // protobuf
+  api 'com.google.protobuf:protobuf-java:3.22.2'
```
Comment (owaiskazi19) [inline server/build.gradle:151]: Same as above

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 7]
```diff
@@ -26,54 +26,54 @@
...
     private static final Logger logger = LogManager.getLogger(ExtensionRequest.class);
-    private final ExtensionsManager.RequestType requestType;
-    private final Optional<String> uniqueId;
+    private final ExtensionRequestProto.ExtensionRequest request;
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:29]: Can we import `ExtensionRequest` like `import org.opensearch.extensions.proto.ExtensionRequestProto.ExtensionRequest`?

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java [hunk 8]
```diff
@@ -26,54 +26,54 @@
...
+    public ExtensionRequest(ExtensionRequestProto.RequestType requestType) {
         this(requestType, null);
     }
 
-    public ExtensionRequest(ExtensionsManager.RequestType requestType, @Nullable String uniqueId) {
-        this.requestType = requestType;
-        this.uniqueId = uniqueId == null ? Optional.empty() : Optional.of(uniqueId);
+    public ExtensionRequest(ExtensionRequestProto.RequestType requestType, @Nullable String uniqueId) {
+        ExtensionRequestProto.ExtensionRequest.Builder builder = ExtensionRequestProto.ExtensionRequest.newBuilder();
+        if (uniqueId != null) {
```
Comment (owaiskazi19) [inline server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:37]: How are we handling the case when `uniqueID` is null?

File: server/src/main/proto/extensions/ExtensionRequestProto.proto [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    REQUEST_EXTENSION_DEPENDENCY_INFORMATION = 5;
+    CREATE_COMPONENT = 6;
+    ON_INDEX_MODULE = 7;
+    GET_SETTINGS = 8;
+}
+
+message ExtensionRequest {
+  RequestType requestType = 1;
+  string uniqueId = 2;
+}
```
Comment (owaiskazi19) [inline server/src/main/proto/extensions/ExtensionRequestProto.proto:32]: This is really amazing!

**Comment by saratvemulapalli** [comment]
Moving to draft while I get back on excluding generated sources for java docs.

File: plugins/repository-hdfs/build.gradle (hunk 4 above)
Comment (saratvemulapalli) [reply plugins/repository-hdfs/build.gradle:120]: Yup we do. We are removing these violations from the plugin.

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java (hunk 7 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:29]: Unfortunately this class is also ExtensionRequest which conflicts. 
We can change the names if you strongly believe. Im ok either way.

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java (hunk 8 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:37]: We dont have to. Protobuf already sets default values[1].

[1] https://protobuf.dev/programming-guides/proto3/#default

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java (hunk 3 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:27]: One step at a time :)
Definitely possible, some more framework to plumb in.

**Comment by saratvemulapalli** [comment]
@reta sorry couldn't get back to this since a week. The change is now ready for review. Everything works :)

File: server/build.gradle [hunk 10]
```diff
@@ -45,6 +48,13 @@ publishing {
 
 archivesBaseName = 'opensearch'
 
+sourceSets {
+  main {
```
Comment (reta) [inline server/build.gradle:52]: 👍

File: server/build.gradle [hunk 11]
```diff
@@ -315,13 +344,34 @@ tasks.named("dependencyLicenses").configure {
     }
 }
 
+tasks.named("missingJavadoc").configure {
```
Comment (reta) [inline server/build.gradle:347]: 👍

**Comment by reta** [comment]
Nice one @saratvemulapalli , pretty clean and simple

File: server/build.gradle [hunk 12]
```diff
@@ -315,13 +344,34 @@ tasks.named("dependencyLicenses").configure {
...
+
 tasks.named("licenseHeaders").configure {
+    dependsOn("generateProto")
     // Ignore our vendored version of Google Guice
     excludes << 'org/opensearch/common/inject/**/*'
     // Ignore temporary copies of impending 8.7 Lucene classes
     excludes << 'org/apache/lucene/search/RegExp87*'
     excludes << 'org/apache/lucene/search/RegexpQuery87*'
     excludes << 'org/opensearch/client/documentation/placeholder.txt'
+    // Ignore for generated code
```
Comment (owaiskazi19) [inline server/build.gradle:373]:
```suggestion
    // Ignore for protobuf generated code
```

File: server/build.gradle [hunk 13]
```diff
@@ -315,13 +344,34 @@ tasks.named("dependencyLicenses").configure {
     }
 }
 
+tasks.named("missingJavadoc").configure {
+  /*
+   * Generated code doesn't follow javadocs formats.
+   * Unfortunately the missingJavadoc task doesnt take *.
+   * TODO: Add support to missingJavadoc task to ignore all generated source code
```
Comment (owaiskazi19) [inline server/build.gradle:351]: Do we have an issue for this?

**Review by owaiskazi19 — COMMENTED** [review]
Overall LGTM! Thanks for addressing the feedback @saratvemulapalli! Do you think we should add tests for the generated Protobuf classes?

File: plugins/repository-hdfs/build.gradle (hunk 4 above)
Comment (owaiskazi19) [reply plugins/repository-hdfs/build.gradle:120]: Sorry if I wasn't clear. I meant do we need `thirdPartyAudit` task here anymore in this build.gradle? Can we remove the task from this build.gradle

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java (hunk 7 above)
Comment (owaiskazi19) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:29]: Ah. I see

File: server/build.gradle (hunk 12 above)
Comment (saratvemulapalli) [reply server/build.gradle]: Sure done.

File: server/build.gradle (hunk 13 above)
Comment (saratvemulapalli) [reply server/build.gradle:351]: Added.

File: server/src/main/java/org/opensearch/extensions/ExtensionRequest.java (hunk 3 above)
Comment (saratvemulapalli) [reply server/src/main/java/org/opensearch/extensions/ExtensionRequest.java:27]: https://stackoverflow.com/questions/41515779/unit-test-proto3-generated-objects-with-verify

**Comment by saratvemulapalli** [comment]
> Overall LGTM! Thanks for addressing the feedback @saratvemulapalli! Do you think we should add tests for the generated Protobuf classes?

Tests for generated classes, I am like "protoc" can you generate unit tests too :D. 
Anyway for generated code its already concise and optimized, no point writing tests for message data types/for internal protobuf business logic.  
@reta WDYT ?

**Comment by saratvemulapalli** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/14394/
> * **CommitID:** [33e500a](https://github.com/opensearch-project/OpenSearch/commit/33e500ad762e4b8a841a76a892b16043dc44e45f)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Looks like 2.7 is on its way, i'll rebase my PR.
```
 [2.6.1] build complete, generating: /var/jenkins/workspace/gradle-check/search/distribution/bwc/bugfix/build/bwc/checkout-2.6/build/14394.tar.bz2
 [2.6.1] 
 [2.6.1] BUILD SUCCESSFUL in 5m 9s
 [2.6.1] 165 actionable tasks: 165 executed

> Task :distribution:bwc:minor:buildBwcLinuxTar
 [2.8.0] > Task :distribution:archives:buildLinuxTar
 [2.8.0] > Task :distribution:archives:linux-tar:assemble
 [2.8.0] build complete, generating: /var/jenkins/workspace/gradle-check/search/distribution/bwc/minor/build/bwc/checkout-2.x/build/14394.tar.bz2
```

File: plugins/repository-hdfs/build.gradle (hunk 4 above)
Comment (saratvemulapalli) [reply plugins/repository-hdfs/build.gradle:120]: Ah I missed that. It makes sense.

**Review by VachaShah — APPROVED** [review]
This looks great!

**Pull request merged**